pkg/controllers/nodeclass/controller.go

```diff
@@ -44,34 +44,103 @@ import (
 	"sigs.k8s.io/karpenter/pkg/events"
 
 	v1 "github.com/aws/karpenter-provider-aws/pkg/apis/v1"
+	"github.com/aws/karpenter-provider-aws/pkg/providers/amifamily"
 	"github.com/aws/karpenter-provider-aws/pkg/providers/instanceprofile"
+	"github.com/aws/karpenter-provider-aws/pkg/providers/launchtemplate"
+	"github.com/aws/karpenter-provider-aws/pkg/providers/securitygroup"
+	"github.com/aws/karpenter-provider-aws/pkg/providers/subnet"
 )
 
+type nodeClassReconciler interface {
+	Reconcile(context.Context, *v1.EC2NodeClass) (reconcile.Result, error)
+}
+
 type Controller struct {
-	kubeClient              client.Client
-	recorder                events.Recorder
-	instanceProfileProvider instanceprofile.Provider
-	launchTemplateProvider  launchtemplate.Provider
+	kubeClient             client.Client
+	recorder               events.Recorder
+	launchTemplateProvider launchtemplate.Provider
+
+	ami             *AMI
+	instanceProfile *InstanceProfile
+	subnet          *Subnet
+	securityGroup   *SecurityGroup
+	validation      *Validation
+	readiness       *Readiness //TODO : Remove this when we have sub status conditions
 }
 
-func NewController(kubeClient client.Client, recorder events.Recorder,
-	instanceProfileProvider instanceprofile.Provider, launchTemplateProvider launchtemplate.Provider) *Controller {
+func NewController(kubeClient client.Client, recorder events.Recorder, subnetProvider subnet.Provider, securityGroupProvider securitygroup.Provider,
+	amiProvider amifamily.Provider, instanceProfileProvider instanceprofile.Provider, launchTemplateProvider launchtemplate.Provider) *Controller {
 
 	return &Controller{
-		kubeClient:              kubeClient,
-		recorder:                recorder,
-		instanceProfileProvider: instanceProfileProvider,
-		launchTemplateProvider:  launchTemplateProvider,
+		kubeClient:             kubeClient,
+		recorder:               recorder,
+		launchTemplateProvider: launchTemplateProvider,
+		ami:                    &AMI{amiProvider: amiProvider},
+		subnet:                 &Subnet{subnetProvider: subnetProvider},
+		securityGroup:          &SecurityGroup{securityGroupProvider: securityGroupProvider},
+		instanceProfile:        &InstanceProfile{instanceProfileProvider: instanceProfileProvider},
+		validation:             &Validation{},
+		readiness:              &Readiness{launchTemplateProvider: launchTemplateProvider},
 	}
 }
 
+func (c *Controller) Name() string {
+	return "nodeclass"
+}
+
 func (c *Controller) Reconcile(ctx context.Context, nodeClass *v1.EC2NodeClass) (reconcile.Result, error) {
-	ctx = injection.WithControllerName(ctx, "nodeclass.termination")
```

Comment:
Seems like we need to call this PR out as a breaking change since these will impact the controller runtime metrics/logs, no? At least call it out as part of our upgrade guide

Reply:
That's a good call out. I was looking at the metrics earlier and it didn't seem like we have metrics specific to each controller: https://karpenter.sh/docs/reference/metrics/ (although someone can still scrape it from Prometheus). Is there a mechanism in place for changes like these? There's no change in behavior or functionality but yes, logging and metrics will change if someone is explicitly relying on metrics for individual controllers.

Reply:
Pushed a new commit to add this callout in the Upgrade Guide.